### PR TITLE
feat: Added store fetch summary routine

### LINKED ISSUE #284: Store Fetch Summary
### Is your feature request related to a problem? Please describe.

It is important to understand what resource(s) were fetched and which resource(s) had issues or warnings. It is also important to understand what provider was used to grab the data. This data can be used during policy evaluation to ensure correctness

### Describe the solution you'd like

At the end of every fetch we should store a summary of all actions that were attempted and their result.

### PR DESCRIPTION
**Pull request by amanenk**
closes #284

### CODE REVIEW

File: pkg/client/fetch_summary/fetch_summary.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+);
+`
+
+type FetchSummarizer struct {
+	pool     *pgxpool.Pool
+	dbStruct *sqlbuilder.Struct
+}
+
+// NewFetchSummarizer creates cq_fetches table and returns a summarizer that saves fetch summary to cq_fetches table
+func NewFetchSummarizer(ctx context.Context, pool *pgxpool.Pool) (*FetchSummarizer, error) {
```
Comment (disq) [inline pkg/client/fetch_summary/fetch_summary.go:39]: just `New` maybe?

File: pkg/client/fetch_summary/fetch_summary.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+	return &FetchSummarizer{
+		pool:     pool,
+		dbStruct: sqlbuilder.NewStruct(new(FetchSummary)),
+	}, nil
+}
+
+// FetchSummary includes a summarized report of fetch, such as fetch id, fetch start and finish,
+// resources fetch results
+type FetchSummary struct {
```
Comment (disq) [inline pkg/client/fetch_summary/fetch_summary.go:59]: `Summary`

File: pkg/client/fetch_summary/fetch_summary.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	Finish             time.Time              `db:"finish"`
+	TotalResourceCount uint64                 `db:"total_resource_count"`
+	ProviderName       string                 `db:"provider_name"`
+	ProviderVersion    string                 `db:"provider_version"`
+	ProviderMeta       []byte                 `db:"provider_meta"` // reserved field to store providers metadata such as
+	FetchedResources   []ResourceFetchSummary `db:"fetch_results"`
+}
+
+// ResourceFetchSummary includes a data about fetching specific resource
+type ResourceFetchSummary struct {
```
Comment (disq) [inline pkg/client/fetch_summary/fetch_summary.go:73]: `ResourceSummary`

File: pkg/client/fetch_summary/fetch_summary.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	Status string `json:"status"`
+	// Total Amount of resources collected by this resource
+	ResourceCount uint64 `json:"resource_count"`
+	// Diagnostics of failed resource fetch, the diagnostic provides insights such as severity, summary and
+	// details on how to solve this issue
+	Diagnostics diag.Diagnostics `json:"diagnostics"`
+}
+
+// SaveFetchSummary saves fetch summary into cq_fetches database
+func (c *FetchSummarizer) SaveFetchSummary(ctx context.Context, fs FetchSummary) error {
```
Comment (disq) [inline pkg/client/fetch_summary/fetch_summary.go:92]: `Save`

File: pkg/client/fetch_summary/fetch_summary.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import (
+	"context"
+	"time"
+
+	"github.com/huandu/go-sqlbuilder"
+	"github.com/jackc/pgx/v4/pgxpool"
+
+	"github.com/cloudquery/cq-provider-sdk/cqproto"
+	"github.com/cloudquery/cq-provider-sdk/provider/schema/diag"
+	"github.com/google/uuid"
```
Comment (disq) [inline pkg/client/fetch_summary/fetch_summary.go:12]: I would move this line to the upper block

**Review by roneli — CHANGES_REQUESTED** [review]
Didn't post this review for some reason, sorry for the delay. Looks like a good start, have some structural changes to purpose  + we need to support migrate for core tables that are created, otherwise it will be hard to manage it.

File: go.mod [hunk 6]
```diff
@@ -133,3 +133,7 @@ require (
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
 )
+
+replace (
+	github.com/cloudquery/cq-provider-sdk => ../cq-provider-sdk
```
Comment (disq) [inline go.mod:138]: go.mod testing mode leak

File: pkg/client/migrations/1_v0.17.4.down.sql [hunk 7]
```diff
@@ -0,0 +1 @@
+CREATE TABLE cq_fetches;
```
Comment (disq) [inline pkg/client/migrations/1_v0.17.4.down.sql:1]:
```suggestion
DROP TABLE cq_fetches;
```

Also drop schema?

File: pkg/client/migrations/1_v0.17.4.up.sql [hunk 8]
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE IF NOT EXISTS cq_fetches
+(
+    cq_id                UUID NOT NULL,
+    cq_fetch_id          UUID NOT NULL,
```
Comment (disq) [inline pkg/client/migrations/1_v0.17.4.up.sql:4]: What's the difference between `cq_id` and `cq_fetch_id`? Also why the `cq_` prefix here?

File: go.mod (hunk 6 above)
Comment (roneli) [reply go.mod:138]: I think we can upgrade to latest SDK @amanenk

File: pkg/client/client.go [hunk 9]
```diff
@@ -291,6 +296,12 @@ func New(ctx context.Context, options ...Option) (*Client, error) {
 			c.Logger.Warn("postgrest validation warning", "err", err)
 		}
 	}
+	// migrate cloudquery core tables to latest version
+	err = c.MigrateCore(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to migrate cloudquery_core tables: %w", err)
+	}
```
Comment (roneli) [inline pkg/client/client.go:303]:
```suggestion
	if err := c.MigrateCore(ctx); err != nil {
		return nil, fmt.Errorf("failed to migrate cloudquery core tables: %w", err)
	}
```

File: pkg/client/migrations/1_v0.17.4.up.sql [hunk 10]
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE IF NOT EXISTS cq_fetches
```
Comment (roneli) [inline pkg/client/migrations/1_v0.17.4.up.sql:1]: we can call this table fetches as users will do cloudquery.fetches to access this table `cq` will be redundant here i.e cloudquery.cq_fetches

File: pkg/client/migrations/1_v0.17.4.up.sql (hunk 8 above)
Comment (roneli) [reply pkg/client/migrations/1_v0.17.4.up.sql:4]: I agree we should remove the cq prefix here, as this table resides in its own schema

File: pkg/client/client.go [hunk 11]
```diff
@@ -922,6 +968,49 @@ func (c *Client) buildProviderMigrator(migrations map[string][]byte, providerNam
 	return m, providerConfig, err
 }
 
+func createCoreSchema(ctx context.Context, pool *pgxpool.Pool) error {
```
Comment (roneli) [inline pkg/client/client.go:971]: move this function down, usually we want to put private functions after public functions for the reader to first only see the exported interface.

**Review by roneli — CHANGES_REQUESTED** [review]
Looking good, had a few extra comments

**Review by disq — APPROVED** [review]

**Review by yevgenypats — DISMISSED** [review]
I would run the migration on `fetch` and not on `new Client` because otherwise we will have to have a database running even for commands that do not require a database such as `cloudquery policy describe` and potentially others.

**Comment by roneli** [comment]
> I would run the migration on `fetch` and not on `new Client` because otherwise we will have to have a database running even for commands that do not require a database such as `cloudquery policy describe` and potentially others.

I would add there are some places that client can be initialized without dsn otherwise it tries to connect, I think also in policy describe it tries. So if dsn != "" its safe to run migrations

File: pkg/client/client.go [hunk 12]
```diff
@@ -1097,3 +1174,46 @@ func reportFetchSummaryErrors(span otrace.Span, fetchSummaries map[string]Provid
...
+	defer conn.Release()
+
+	_, err = conn.Exec(ctx, "CREATE SCHEMA IF NOT EXISTS cloudquery")
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func readMigrationFiles(log hclog.Logger, migrationFiles embed.FS) (map[string][]byte, error) {
```
Comment (roneli) [inline pkg/client/client.go:1192]: why not use SDK's function?

File: pkg/client/client.go [hunk 13]
```diff
@@ -944,6 +992,35 @@ func (c *Client) buildProviderMigrator(migrations map[string][]byte, providerNam
...
+
+	defer func() {
+		if err := m.Close(); err != nil {
+			c.Logger.Error("failed to close migrator connection", "error", err)
+		}
+	}()
+
+	version := "latest"
+	err = m.UpgradeProvider(version)
+	if err != migrate.ErrNoChange {
```
Comment (roneli) [inline pkg/client/client.go:1018]:
```suggestion
        const latestVersion := "latest" // todo move this out of the function
	if err := m.UpgradeProvider(latestVersion); err != migrate.ErrNoChange {
```

File: pkg/client/client.go [hunk 14]
```diff
@@ -944,6 +992,35 @@ func (c *Client) buildProviderMigrator(migrations map[string][]byte, providerNam
...
+		return err
```
Comment (roneli) [inline pkg/client/client.go:1019]:
```suggestion
		return fmt.Errorf("failed to migrate cloudquery core schema: %w",  err)
```

**Review by roneli — CHANGES_REQUESTED** [review]
More comments I missed in prior review

**Review by roneli — CHANGES_REQUESTED** [review]
Another comment I missed

File: pkg/client/client.go (hunk 12 above)
Comment (amanenk) [reply pkg/client/client.go:1191]: it was not released

**Comment by amanenk** [comment]
> > I would run the migration on `fetch` and not on `new Client` because otherwise we will have to have a database running even for commands that do not require a database such as `cloudquery policy describe` and potentially others.
> 
> I would add there are some places that client can be initialized without dsn otherwise it tries to connect, I think also in policy describe it tries. So if dsn != "" its safe to run migrations

If we will have some other core tables related to other commands it is better to migrate in the client initialization

File: pkg/client/client.go [hunk 15]
```diff
@@ -512,6 +551,16 @@ func (c *Client) Fetch(ctx context.Context, request FetchRequest) (res *FetchRes
...
+
+				fs.addResourceFetchSummary(ResourceFetchSummary{
+					ResourceName:                resp.ResourceName,
+					FinishedResources:           resp.FinishedResources,
+					Status:                      strconv.Itoa(int(resp.Summary.Status)), // todo use human readable representation of status
+					Error:                       resp.Error,
+					PartialFetchFailedResources: resp.PartialFetchFailedResources,
+					ResourceCount:               resp.ResourceCount,
+					Diagnostics:                 resp.Summary.Diagnostics,
+				})
```
Comment (roneli) [inline pkg/client/client.go:563]:
```suggestion
				fs.Resources = append(fs.Resources, ResourceFetchSummary{
					ResourceName:                resp.ResourceName,
					FinishedResources:           resp.FinishedResources,
					Status:                      strconv.Itoa(int(resp.Summary.Status)), // todo use human readable representation of status
					Error:                       resp.Error,
					PartialFetchFailedResources: resp.PartialFetchFailedResources,
					ResourceCount:               resp.ResourceCount,
					Diagnostics:                 resp.Summary.Diagnostics,
				})
```

File: pkg/client/fetch.go [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+	//  Unique Id of fetch session
+	FetchId            uuid.UUID               `db:"fetch_id"`
+	Start              time.Time               `db:"start"`
+	Finish             time.Time               `db:"finish"`
+	IsSuccess          bool                    `db:"is_success"`
+	TotalResourceCount uint64                  `db:"total_resource_count"`
+	TotalErrorsCount   uint64                  `db:"total_errors_count"`
+	ProviderName       string                  `db:"provider_name"`
+	ProviderVersion    string                  `db:"provider_version"`
+	FetchedResources   *[]ResourceFetchSummary `db:"results"`
```
Comment (roneli) [inline pkg/client/fetch.go:29]:
```suggestion
	Resources   *[]ResourceFetchSummary `db:"results"`
```

**Review by roneli — APPROVED** [review]
LGTM, two minor nits

**Pull request merged**